Fix the `<>;` redirection operator in `-c` scripts

JohnoKing opened the pull request:
This bug was first reported at https://github.com/att/ast/issues/9.

The `<>;` operator doesn't work correctly if it's used in the last command of a `-c` script. Reproducer:
```sh
$ echo test > a; ksh -c 'echo x 1<>; a'; cat a
x
st
```
This bug is caused by ksh running the last command of `-c` scripts with `execve(2)` instead of `posix_spawn(3)` or `fork(2)`. The `<>;` operator is noted by the man page as being incompatible with the `exec` builtin (see also the ksh93u+ man page), so it's not surprising this bug occurs when ksh runs a command using `execve`:
> <>;word cannot be used with the exec and redirect built-ins.

The ksh2020 fix simply removed the code required for ksh to use this optimization at all, specifically this block:
https://github.com/ksh93/ksh/blob/519bb082651390239040005cffbcbed5075d1e4f/src/cmd/ksh93/sh/main.c#L577-L583
That's not a performance friendly fix and only papers over the bug, so this pull request provides a better fix.

`src/cmd/ksh93/sh/xec.c`:
\- The `IOREWRITE` flag is set when handling the `<>;` operator, so to fix this bug avoid exec'ing the last command if it uses `<>;`, because that redirection operator is inherently incompatible with `exec`. See also commit 17ebfbf6, which fixed another issue related to the `execve` optimization.

`src/cmd/ksh93/tests/io.sh`:
\- Enable a regression test that was previously failing because of this bug.
\- Add the reproducer from https://github.com/att/ast/issues/9 as a regression test.

`src/cmd/ksh93/tests/options.sh`:
\- This bugfix was triggering a segfault in the options regression tests when run under shcomp. The cause is the same as https://github.com/ksh93/ksh/issues/165, so as a workaround avoid parsing process substitutions with shcomp if used with a redirection. This workaround should also avoid the other problem detailed in https://github.com/ksh93/ksh/issues/274.

JohnoKing committed:
Fix the <>; redirection operator in -c scripts

This bug was first reported at https://github.com/att/ast/issues/9.

The <>; operator doesn't work correctly if it's used as the last command
of a -c script. Reproducer:
  $ echo test > a; ksh -c 'echo x 1<>; a'; cat a
  x
  st
This bug is caused by ksh running the last command of -c scripts with
execve(2) instead of posix_spawn(3) or fork(2). The <>; operator is
noted by the man page as being incompatible with the exec builtin (see
also the ksh93u+ man page), so it's not surprising this bug occurs
when ksh runs a command using execve:

> <>;word cannot be used with the exec and redirect built-ins.

The ksh2020 fix simply removed the code required for ksh to use this
optimization at all. It's not a performance friendly fix and only papers
over the bug, so this commit provides a better fix.

src/cmd/ksh93/sh/xec.c:
- The IOREWRITE flag is set when handling the <>; operator, so to fix
  this bug avoid exec'ing the last command if it uses <>;. See also
  commit 17ebfbf6, which fixed another issue related to the execve
  optimization.

src/cmd/ksh93/tests/io.sh:
- Enable a regression test that was failing because of this bug.
- Add the reproducer from https://github.com/att/ast/issues/9 as a
  regression test.

src/cmd/ksh93/tests/options.sh:
- This bugfix was causing the options regression test to segfault when
  run under shcomp. The cause is the same as https://github.com/ksh93/ksh/issues/165,
  so as a workaround avoid parsing the command substitution with shcomp.
  This workaround should also avoid the other problem detailed in
  https://github.com/ksh93/ksh/issues/274.

JohnoKing committed:
To workaround another segfault caused by #165, avoid procsubs here also

McDutchie commented on src/cmd/ksh93/tests/builtins.sh:
```diff
@@ -986,7 +986,10 @@ EOF
+# Note: To workaround https://github.com/ksh93/ksh/issues/165, put the list
+# of builtins in a file, then read from that.
+builtin > "$tmp/builtin-list"
+while IFS= read -r bltin
```
Some built-ins may read from standard input, so this test should continue to use file descriptor 3 (for the `read` and the input redirection following `done`).

McDutchie commented on src/cmd/ksh93/tests/options.sh:
```diff
@@ -537,8 +537,12 @@ fi # SHOPT_BRACEPAT
+
+	# Note: To workaround erratic behavior caused by https://github.com/ksh93/ksh/issues/165,
+	# avoid parsing the process substitutions with shcomp by running the
+	# whole command with ksh -c.
```
Running it with `eval` should work as well, but is faster.

JohnoKing committed:
Use file descriptor 3

JohnoKing committed:
Use eval for faster delayed parsing

JohnoKing committed:
Enable execve optimization in normal scripts

src/cmd/ksh93/sh/main.c:
- Only avoid the non-forking optimization in interactive shells.

src/cmd/ksh93/tests/signal.sh:
- Add an extra comment to avoid the non-forking optimization in the
  regression test for rhbz#1469624.
- If the regression test for rhbz#1469624 fails, show the incorrect
  exit status in the error message.

JohnoKing committed:
Add parentheses
